### ISSUE TITLE
Markdown header > color entire line

### ISSUE THREAD

**Issue by cdbdev** [ScintillaOrg/lexilla#60]
Hi,

In case of a header, only the '#' character has a specific color, but not the entire header line. 

Below an example:

```
# Header 1  

Test
```

I've posted this question already in 2020, but at that time there was nobody maintaining the markdown lexer, until now i guess :)
Is it possible to have a look at this?

Thank you

**Comment by RaiKoHoff** [comment]
Hello,
I changed the original Markdown Lexer targeting this issue here - so that it works for me.
See https://github.com/rizonesoft/Notepad3/blob/master/lexilla/lexers_x/LexMarkdown.cxx.
But these changes have neither been reviewed by someone else nor have they been **_thoroughly_** tested.
Best regards.

![image](https://user-images.githubusercontent.com/8806350/154988714-20c2b92f-668f-46bf-b6a7-44207b1e3cad.png)

**Comment by nyamatongwe** [comment]
Lexers may have properties that can be set by applications to change behaviour. See LexTeX.cxx for examples of `GetPropertyInt` and use a `lexer.markdown.` prefix on the property name. The formatted comments before the call should also be mimicked as they are scraped for code generation and documentation.

**Comment by rdipardo** [comment]
@cdbdev,
> I am willing to test this myself, but I don't know the correct procedure to add a custom lexer..

Assuming you're on Windows, here's a script that will build SciTE with @RaiKoHoff's Markdown lexer.
(Hint &#x2014; use the "copy" button that GitHub puts in the top right corner of code fences):

```bat
@echo off
SETLOCAL
:: Must be run from a Developer Command Prompt, e.g.
:: "C:\Program Files (x86)\Microsoft Visual Studio\2019\Community\Common7\Tools\VsDevCmd.bat"
::
:: Requirements
:: ------------
:: * git <https://gitforwindows.org>
:: * mercurial <https://www.mercurial*scm.org>
:: * GNU sed <https://community.chocolatey.org/packages/sed> -- with WSL installed, `wsl sed <cmd>` will work
:: * cmake <https://cmake.org>
:: * python <https://www.python.org/downloads/>
:: * Visual Studio (just the Build Tools is enough): https://docs.microsoft.com/en*us/cpp/build/vscpp*step*0*installation
if not exist lexilla-dev (
  git clone -b scite-gtk3 --recursive https://bitbucket.org/rdipardo/lexilla-dev.git
  cd lexilla-dev
  git submodule update --init --remote

  if not exist scintilla (hg clone http://hg.code.sf.net/p/scintilla/code scintilla)

  curl -s https://raw.githubusercontent.com/rizonesoft/Notepad3/master/lexilla/lexers_x/LexMarkdown.cxx -o lexilla\lexers\LexMarkdown.cxx
  curl -s https://raw.githubusercontent.com/rizonesoft/Notepad3/master/lexilla/include/LexicalStyles.iface -o lexilla\include\LexicalStyles.iface

  cd lexilla\scripts
  python3 LexFacer.py
  cd ..\..

  if not exist scite (hg clone http://hg.code.sf.net/p/scintilla/scite)
)

cd lexilla-dev
git reset --hard scite-gtk3
git checkout scite-gtk3
dir /b

wsl sed --version >NUL
if %errorlevel% == 0 (SET "SED_CMD=wsl sed") else (SET "SED_CMD=sed")

:: remove setting that excludes Markdown from the Language menu, since properties are compiled with EXE on Windows
%SED_CMD% -i '/^\(markdown \).*\\/d' scite/src/Embedded.properties
.\build.bat
cd ..
ENDLOCAL
```

![lmarkdown_mod3](https://user-images.githubusercontent.com/59004801/155045270-da73116c-29a5-4fc0-b0e7-089234c520f7.png)

**Comment by zufuliu** [comment]
Here is a more feature complete Markdown lexer (for [GitHub Flavored Markdown](https://github.github.com/gfm/), [GitLab Flavored Markdown](https://docs.gitlab.com/ee/user/markdown.html) and [Pandoc’s Markdown](https://pandoc.org/MANUAL.html#pandocs-markdown)), it can be ported to lexilla after issue #54 is done.
https://github.com/zufuliu/notepad2/blob/main/scintilla/lexers/LexMarkdown.cxx

![Notepad2's Markdown](https://user-images.githubusercontent.com/2289926/152997007-8d50fdd7-b336-4d04-843e-6a40b72e6c48.png)

**Comment by cdbdev** [comment]
> @cdbdev,
> 
> > I am willing to test this myself, but I don't know the correct procedure to add a custom lexer..
> 
> Assuming you're on Windows, here's a script that will build SciTE with @RaiKoHoff's Markdown lexer. (Hint — use the "copy" button that GitHub puts in the top right corner of code fences):
> 
> ```batchfile
> @echo off
> SETLOCAL
> :: Must be run from a Developer Command Prompt, e.g.
> :: "C:\Program Files (x86)\Microsoft Visual Studio\2019\Community\Common7\Tools\VsDevCmd.bat"
> ::
> :: Requirements
> :: ------------
> :: * git <https://gitforwindows.org>
> :: * mercurial <https://www.mercurial*scm.org>
> :: * GNU sed <https://community.chocolatey.org/packages/sed> -- with WSL installed, `wsl sed <cmd>` will work
> :: * cmake <https://cmake.org>
> :: * python <https://www.python.org/downloads/>
> :: * Visual Studio (just the Build Tools is enough): https://docs.microsoft.com/en*us/cpp/build/vscpp*step*0*installation
> if not exist lexilla-dev (
>   git clone -b scite-gtk3 --recursive https://bitbucket.org/rdipardo/lexilla-dev.git
>   cd lexilla-dev
>   git submodule update --init --remote
> 
>   if not exist scintilla (hg clone http://hg.code.sf.net/p/scintilla/code scintilla)
> 
>   curl -s https://raw.githubusercontent.com/rizonesoft/Notepad3/master/lexilla/lexers_x/LexMarkdown.cxx -o lexilla\lexers\LexMarkdown.cxx
>   curl -s https://raw.githubusercontent.com/rizonesoft/Notepad3/master/lexilla/include/LexicalStyles.iface -o lexilla\include\LexicalStyles.iface
> 
>   cd lexilla\scripts
>   python3 LexFacer.py
>   cd ..\..
> 
>   if not exist scite (hg clone http://hg.code.sf.net/p/scintilla/scite)
> )
> 
> cd lexilla-dev
> git reset --hard scite-gtk3
> git checkout scite-gtk3
> dir /b
> 
> wsl sed --version >NUL
> if %errorlevel% == 0 (SET "SED_CMD=wsl sed") else (SET "SED_CMD=sed")
> 
> :: remove setting that excludes Markdown from the Language menu, since properties are compiled with EXE on Windows
> %SED_CMD% -i '/^\(markdown \).*\\/d' scite/src/Embedded.properties
> .\build.bat
> cd ..
> ENDLOCAL
> ```
> 
> ![lmarkdown_mod3](https://user-images.githubusercontent.com/59004801/155045270-da73116c-29a5-4fc0-b0e7-089234c520f7.png)

Hi,

Is it possible to compile 64bit with msvc x86_64? Because I always get the message c:\program files (x86) ... Not Found message.

Thanks

**Comment by rdipardo** [comment]
If you've cloned the repo, look at `build.bat` and change the platform option at line 16 from "x86" to "x64"

```diff
-msbuild /nologo /v:m /p:Platform=x86;WindowsTargetPlatformVersion=%WindowsSDKVersion:~0,-1%
+msbuild /nologo /v:m /p:Platform=x64;WindowsTargetPlatformVersion=%WindowsSDKVersion:~0,-1%
```

I should have also mentioned that cmake isn't actually needed to build SciTE on Windows (the *.vcxproj descriptor pulls everything in by itself).

**Comment by cdbdev** [comment]
> If you've cloned the repo, look at `build.bat` and change the platform option at line 16 from "x86" to "x64"
> 
> ```diff
> -msbuild /nologo /v:m /p:Platform=x86;WindowsTargetPlatformVersion=%WindowsSDKVersion:~0,-1%
> +msbuild /nologo /v:m /p:Platform=x64;WindowsTargetPlatformVersion=%WindowsSDKVersion:~0,-1%
> ```
> 
> I should have also mentioned that cmake isn't actually needed to build SciTE on Windows (the *.vcxproj descriptor pulls everything in by itself).

Doesn't seem to work, I get the following error:

```
C:\Program Files (x86)\Microsoft Visual Studio\2019\BuildTools\MSBuild\Microsoft\VC\v160\Microsoft.CppBuild.targets(439
,5): error MSB8020: The build tools for Visual Studio 2017 (Platform Toolset = 'v141') cannot be found. To build using
the v141 build tools, please install Visual Studio 2017 build tools.  Alternatively, you may upgrade to the current Vis
ual Studio tools by selecting the Project menu or right-click the solution, and then selecting "Retarget solution". [c:
\dev\scite-lexilla\lexilla-dev\scite\win32\SciTE.vcxproj]
```
I'm running VS 2019.

**Comment by cdbdev** [comment]
I get the following errors during compilation:

```
LexMarkdown.cxx(121,35): error C2065: 'SCE_MARKDOWN_HDRTEXT': undeclare
d identifier [c:\dev\scite-lexilla\lexilla-dev\scite\win32\SciTE.vcxproj]
...
'scite\win32\Debug\SciTE.exe' is not recognized as an internal or external command,
operable program or batch file.
```

**Comment by rdipardo** [comment]
> ```
> LexMarkdown.cxx(121,35): error C2065: 'SCE_MARKDOWN_HDRTEXT': undeclare
> d identifier [c:\dev\scite-lexilla\lexilla-dev\scite\win32\SciTE.vcxproj]
> ```

This means the LexFacer.py script did not run successfuly.
From the `lexilla/scripts` directory, run `python3 LexFacer.py`; the output should be like this:

    Changed C:\Users\Rob\workspace\lexilla-dev\lexilla\include\SciLexer.h

> ```
> 'scite\win32\Debug\SciTE.exe' is not recognized as an internal or external command,
> operable program or batch file.
> ```

If you built for x64, the app will be at "scite\win32\x64\Debug\SciTE.exe":

```diff
-scite\win32\Debug\SciTE.exe
+scite\win32\x64\Debug\SciTE.exe
```

**Comment by cdbdev** [comment]
Sorry for the spamming.

The build worked fine now, but I do not get any syntax coloring form markdown. I see that there is no markdown.properties file in the menu either. And the global options is also empty, so from what I can tell this is a "virgin" version of Scite?

Also, were can I find the generated DLL?

Thanks for all your help!

**Comment by rdipardo** [comment]
> is it also possible to generate a separate lexilla.dll to use with the latest scite, or is this not possible?
> 

The Visual C++ project doesn't produce a DLL because it [statically links] all the components.

You can build a shared lexilla library using the `build-nmake.bat` script in the development repo, which does however depend on cmake. The default target is a test program that will fail on the custom lexer (and delete the library), so check the  [usage notes] and make the appropriate edits, e,g,:

```diff
-cmake -Bbuild-msvc -H. -G"NMake Makefiles" -DLEXILLA_STATIC=1 -DDEBUG=1
-cmake --build ./build-msvc -- TestLexers
+cmake -Bbuild-msvc -H. -G"NMake Makefiles" -DDEBUG=1
+cmake --build ./build-msvc -- lexilla
```

[statically links]: https://sourceforge.net/p/scintilla/scite/ci/default/tree/win32/SciTE.vcxproj
[usage notes]: https://bitbucket.org/rdipardo/lexilla-dev/src/default/README.md

**Comment by nyamatongwe** [comment]
The Scintilla mailing list is used for correspondence about lexers.
https://groups.google.com/forum/#!forum/scintilla-interest

This change does not provide an option for the previous behaviour. Unless there was consensus from all involved that the new behaviour is better then it should be optional.

**Comment by cdbdev** [comment]
> The Scintilla mailing list is used for correspondence about lexers. https://groups.google.com/forum/#!forum/scintilla-interest
> 
> This change does not provide an option for the previous behaviour. Unless there was consensus from all involved that the new behaviour is better then it should be optional.

I'm afraid I don't understand. The solution works, the entire line is now highlighted instead of just the # char. Or is it because the proposed solution does not correspond to the working of other lexers?

**Comment by nyamatongwe** [comment]
>  the entire line is now highlighted instead of just the # char

This is an unconditional change in behaviour. Some may prefer this behaviour and others may prefer the original behaviour of only highlighting the '#'s. The original lexer author could have implemented headers in this way but did not.

The change adds a new SCE_MARKDOWN_HDRTEXT style but there is no explanation of its purpose. The change has complicated the use of the `freezeCursor` variable in a way that makes no sense to me.

There is no clear applicable patch or pull request.

**Comment by zufuliu** [comment]
(Just mt guess as Notepad2 use my own lexer) I think original lexer author not implemented this is want to keep inline formatting work inside headers, like following (for https://github.com/microsoft/TypeScript/blob/main/doc/spec-ARCHIVED.md):

![Notepad2-TypeScript-Spec](https://user-images.githubusercontent.com/2289926/155861654-285f411c-0162-4f88-bec7-22a21a27abce.png)

**Comment by nyamatongwe** [comment]
Worked out `SCE_MARKDOWN_HDRTEXT`. Its possible to use more '#'s to **[*close*](https://daringfireball.net/projects/markdown/syntax#header)** a header and text after the close is in `SCE_MARKDOWN_HDRTEXT`. However, the lexer has no way to end the `SCE_MARKDOWN_HDRTEXT` so it just continues to the end of file.
~~~markdown
# Heading level 1=6 # Some SCE_MARKDOWN_HDRTEXT
~~~
![MarkdownHDRTEXT](https://user-images.githubusercontent.com/225818/155864520-b24484d7-3813-484a-8184-10870da4138e.png)

This interpretation of '#' conflicts with [Heading IDs](https://www.markdownguide.org/extended-syntax/#heading-ids).
~~~markdown
### My Great Heading {#custom-id}
~~~

**Comment by cdbdev** [comment]
Ok thank you. It makes sense now. I understand why it should be made configurable.

Hopefully someone can have a look :)

**Comment by nyamatongwe** [comment]
> text after the close is in `SCE_MARKDOWN_HDRTEXT`

From the CommonMark specification, any (non-whitespace) text following the header close is part of the header.
https://spec.commonmark.org/current/#atx-headings

**Comment by cdbdev** [comment]
Wouldn't it be an option to work out the `SetStateAndZoom` method a bit more so it handles a specific property (when provided) for highlighting the entire line. Somewhat like you suggested in your first post on this issue?

**Comment by cdbdev** [comment]
I'm trying out some coding myself, but I can't get the property to work.

I added the following property to the end of the `markdown.properties` file:

```
lexer.markdown.header.eolfill=1
```

But when I try to read it in the code, it does not work:

```
bool headerEOLFill   = styler.GetPropertyInt("lexer.markdown.header.eolfill") == 1 ;
```
I'm probably forgetting something...

Can anyone help me, thanks! :)

**Comment by cdbdev** [comment]
I've made some changes to the original version. If someone wants to review, please have a look here:
[LexMarkdown.cxx](https://github.com/cdbdev/lexilla/blob/master/lexers/LexMarkdown.cxx)

Basically I just check if the new property `lexer.markdown.header.eolfill` is equal to 1. If this is the case, the entire line is highlighted. If not, the original logic is used.

**Comment by nyamatongwe** [comment]
It is not quite the same effect as the first image above. There the background colour can fill the whole line because it styles the new line characters then uses an `eolfilled` style. Ending the style before the line end may be fine but I'd expect some people to want to be able to choose the illustrated effect.

There should be a formatted comment before the `GetPropertyInt ` call like in other lexers as this is used for documentation and UI. There's not much point to properties unless they can be discovered.

**Comment by cdbdev** [comment]
What do you mean exactly with **formatted comment**? If I look at LexTex, I don't see it:

```
styler.StartAt(startPos) ;
styler.StartSegment(startPos) ;

bool processComment   = styler.GetPropertyInt("lexer.tex.comment.process",   0) == 1 ;
```

**Comment by cdbdev** [comment]
I'm trying to mimic the eolfill from the first screenshot, but can't get it to work without filling the line with spaces. Is this a specific setting I'm missing?

**Comment by cdbdev** [comment]
I got it working now with eolfilled:
[LexMarkdown.cxx](https://github.com/cdbdev/lexilla/blob/master/lexers/LexMarkdown.cxx)

Can  you have a second look?

Thanks

**Comment by nyamatongwe** [comment]
The updated code is applying the style too far and missing the initial `#` on the next line.
![MarkdownHeader2](https://user-images.githubusercontent.com/225818/156463470-fe25ff69-ee4a-490d-8372-ca1917bf5950.png)

It may be simpler to pass `headerEOLFill` directly into `SetStateAndZoom` as a distinct parameter rather than fiddling the `length` to a magic number.

The code has been reformatted which makes it more difficult to work out what was changed. `LexMarkdown.cxx` is formatted unusually but that should be respected and continued to minimize change sets and ease integration of contributions.

A short test case should be written for the feature. This will ensure that the behaviour is checked when other changes are made and that it works with both Windows and Unix line ends. As this feature is changing a property, the test should be in a subdirectory of `lexilla/test/examples/markdown`. See `lexilla/test/README` for information on testing.

**Comment by mpheath** [comment]
I have added underlined headers to the *AllStyles.md* test file as were markdown items missing and is related to *H1* and *H2* headers styling, in my opinion. *AllStyles.md* passes the test before and after the update of the test file.

Setting *H1* or *H2* with back colour and eolfilled causes the underlined header text to have mixed *EOL* stying and can back colour the whole line when it should not as it affects the *LF*, so a small modification to `FollowToLineEnd()` fixes this minor issue.

The main `while` loop in *LexMarkdown.cxx* has some header handling code so I modified that and changed code in the functions if needed, rather than trying to make function code handle the headers and hope that the code in the `while` loop agrees with it.

Pretty fresh though seems to be OK. The headers work with or without surrounding empty lines even though it is usually recommended to have the empty lines.

This image displays additional `eolfilled,back:#FFFF99` styling to the default *H1* and *H2* header styles.

![LexMarkdown](https://user-images.githubusercontent.com/58158242/156751821-e2732af3-f2e5-482b-a601-9f46719ae66d.gif)

Patch file:
[LexMarkdown.zip](https://github.com/ScintillaOrg/lexilla/files/8185295/LexMarkdown.zip)

If @cdbdev has something better, then that will be OK to use instead. ;)

**Comment by cdbdev** [comment]
I don't think I've got anything better :D. 
Didn't even touch the header underline issue.

Ok for me, but waiting for response of @nyamatongwe :D

**Comment by nyamatongwe** [comment]
@mpheath, that looks good with `lexer.markdown.header.eolfill` both on and off.

A test case should be added for `lexer.markdown.header.eolfill=1`.

**Comment by mpheath** [comment]
Glad it is confirmed as good. Here are the new subdirectory test files, both LF and CRLF. Take your pick.

[TestMarkdown.zip](https://github.com/ScintillaOrg/lexilla/files/8190076/TestMarkdown.zip)

The `lexer.markdown.header.eolfill=1` line in the new *SciTE.properties* had a previous EOL sequence of only CR and caused the property not to be read with running *TestLexer.exe*. Do not know how the CR got there and made me look high and low everywhere else, until I eventually found it by displaying EOLs.

**Comment by cdbdev** [comment]
Great news, thanks :)

Just 2 questions:

- Will this functionality be in one of the next releases if possible?
- can markdown be removed from the excluded list in the global options?

**Comment by nyamatongwe** [comment]
@mpheath:
> had a previous EOL sequence of only CR and caused the property not to be read

The simple properties reader in TestLexers uses `std::getline` so treats a bare CR as not ending a line. The returned value includes the first line, the CR, and the next line. With an embedded CR, that will parse poorly. These files will have native line ends when downloaded from git or from an archive so it's unlikely this will be a problem. It doesn't appear worth rewriting `PropertyMap::ReadFromFile`.

**Comment by nyamatongwe** [comment]
@cdbdev:
It's likely that this will be included in a near future release of SciTE. When it goes in depends on why the SciTE release is made and the apparent stability of Lexilla at that point. SciTE 5.2.1 was an urgent bug fix so Lexilla wasn't updated then.

> can markdown be removed from the excluded list in the global options?

Pick a language to retire from inclusion and make your case on the SciTE mailing list.

**Comment by nyamatongwe** [comment]
#62 defines a way to avoid extra subdirectories for testing properties in simple cases. That technique will be used for the `lexer.markdown.header.eolfill=1` test so this issue now depends on #62.

**Comment by mpheath** [comment]
> #62 defines a way to avoid extra subdirectories for testing properties in simple cases. That technique will be used for the `lexer.markdown.header.eolfill=1` test so this issue now depends on #62.

New test files based on new technique:

[markdown.zip](https://github.com/ScintillaOrg/lexilla/files/8192247/markdown.zip)

Added 2 tests this time for both `lexer.markdown.header.eolfill=0` and `lexer.markdown.header.eolfill=1`. Markdown has no comment character so used backquotes to enclose the embedded properties.

**Comment by nyamatongwe** [comment]
> Markdown has no comment character

I hadn't considered that. Quoting with `` ` `` appears reasonable for Markdown.

**Comment by cdbdev** [comment]
I believe you can use html like comment:

```
<!---
your comment ...
-->
```

**Comment by nyamatongwe** [comment]
The feature has now been committed.

`cdbdev` was added to the credits. If you want a different name then tell me what you would prefer.

Testing of `lexer.markdown.header.eolfill` was enabled by #62 Allows setting per-file properties in SciTE.properties.

**Comment by cdbdev** [comment]
Super, thank you!

cdbdev is fine :)

**Comment by cdbdev** [comment]
Available in release 5.2.2.
